block_id.gen.go

```diff
@@ -0,0 +1,201 @@
+)
+
+var (
+	ErrInvalidBlockIDLength = ierrors.New("invalid blockID length")
+
+	EmptyBlockID = BlockID{}
+)
+
+// BlockID is a 32 byte hash value together with an 4 byte slot index.
+type BlockID [BlockIDLength]byte
+
+// BlockIDRepresentingData returns a new BlockID for the given data by hashing it with blake2b and associating it with the given slot index.
+func BlockIDRepresentingData(slot SlotIndex, data []byte) BlockID {
+	return NewBlockID(slot, blake2b.Sum256(data))
+}
+
+func NewBlockID(slot SlotIndex, idBytes Identifier) BlockID {
+	b := BlockID{}
+	copy(b[:], idBytes[:])
+	binary.LittleEndian.PutUint32(b[IdentifierLength:], uint32(slot))
+
+	return b
+}
+
+// BlockIDFromHexString converts the hex to a BlockID representation.
+func BlockIDFromHexString(hex string) (BlockID, error) {
+	b, err := hexutil.DecodeHex(hex)
+	if err != nil {
+		return BlockID{}, err
+	}
+
+	s, _, err := BlockIDFromBytes(b)
+
+	return s, err
+}
+
+// BlockIDFromBytes returns a new BlockID represented by the passed bytes.
+func BlockIDFromBytes(b []byte) (BlockID, int, error) {
+	if len(b) < BlockIDLength {
+		return BlockID{}, 0, ErrInvalidBlockIDLength
+	}
+
+	return BlockID(b), BlockIDLength, nil
+}
+
+// MustBlockIDFromHexString converts the hex to a BlockID representation.
+func MustBlockIDFromHexString(hex string) BlockID {
+	b, err := BlockIDFromHexString(hex)
+	if err != nil {
+		panic(err)
+	}
+
+	return b
+}
+
+func (b BlockID) Bytes() ([]byte, error) {
+	return b[:], nil
+}
+
+func (b BlockID) MarshalText() (text []byte, err error) {
+	dst := make([]byte, hex.EncodedLen(len(BlockID{})))
+	hex.Encode(dst, b[:])
+
+	return dst, nil
+}
+
+func (b *BlockID) UnmarshalText(text []byte) error {
+	_, err := hex.Decode(b[:], text)
+
+	return err
+}
+
+// Empty tells whether the BlockID is empty.
+func (b BlockID) Empty() bool {
+	return b == EmptyBlockID
+}
+
+// ToHex converts the Identifier to its hex representation.
+func (b BlockID) ToHex() string {
+	return hexutil.EncodeHex(b[:])
+}
+
+func (b BlockID) String() string {
+	return fmt.Sprintf("BlockID(%s:%d)", b.Alias(), b.Slot())
+}
+
+func (b BlockID) Slot() SlotIndex {
+	return SlotIndex(binary.LittleEndian.Uint32(b[IdentifierLength:]))
+}
+
+// Index returns a slot index to conform with hive's IndexedID interface.
+func (b BlockID) Index() SlotIndex {
+	return b.Slot()
+}
+
+func (b BlockID) Identifier() Identifier {
+	return Identifier(b[:IdentifierLength])
+}
+
+var (
+	// BlockIDAliases contains a dictionary of identifiers associated to their human-readable alias.
+	BlockIDAliases = make(map[BlockID]string)
+
+	// BlockIDAliasesMutex is the mutex that is used to synchronize access to the previous map.
+	BlockIDAliasesMutex = sync.RWMutex{}
+)
+
+// RegisterAlias allows to register a human-readable alias for the Identifier which will be used as a replacement for
+// the String method.
+func (b BlockID) RegisterAlias(alias string) {
+	BlockIDAliasesMutex.Lock()
```

Comment:
Seems like the "firstLower .Name" doesn't work correctly?

Reply:
It works but I didn't call it for the `slot_identifier` template